File: ios/NorthAx/Models/ActivityStreams.swift
import Foundation

/// Downsampled time-series for a completed activity (§10). Arrays are
/// index-aligned with `time` (seconds from start); any absent metric is empty.
struct ActivityStreams {
    var activityId: String
    var time: [Double]
    var heartRate: [Double]
    var power: [Double]
    var velocity: [Double]   // m/s
    var altitude: [Double]
    var cadence: [Double]
    var source: String

    var hasData: Bool {
        !heartRate.isEmpty || !power.isEmpty || !velocity.isEmpty
            || !altitude.isEmpty || !cadence.isEmpty
    }

    /// Speed in km/h (from m/s) — intuitive for a review chart (higher = faster).
    var speedKmh: [Double] { velocity.map { $0 * 3.6 } }

    var durationSeconds: Double { time.last ?? 0 }
}

File: android/app/src/main/java/app/northax/domain/model/ActivityStreams.kt
package app.northax.domain.model

/**
 * Downsampled time-series for a completed activity. Arrays are index-aligned
 * with `time` (seconds from start); any absent metric is empty.
 */
data class ActivityStreams(
    val activityId: String,
    val time: List<Double>,
    val heartRate: List<Double>,
    val power: List<Double>,
    val velocity: List<Double>, // m/s
    val altitude: List<Double>,
    val cadence: List<Double>,
    /**
     * GPS route as [[lat, lng], …]; denser than the scalar arrays and NOT
     * index-aligned with `time`. Empty for indoor/virtual activities.
     */
    val latLng: List<List<Double>> = emptyList(),
    val source: String,
) {
    val hasData: Boolean
        get() = heartRate.isNotEmpty() || power.isNotEmpty() || velocity.isNotEmpty() ||
            altitude.isNotEmpty() || cadence.isNotEmpty()

    /** Speed in km/h (from m/s) — intuitive for a review chart. */
    val speedKmh: List<Double> get() = velocity.map { it * 3.6 }

    val durationSeconds: Double get() = time.lastOrNull() ?: 0.0
}
